### PR TITLE
[Chess] Reimplement from scratch

### PR DESCRIPTION
**Pull request by sotetsuk**
- [x] step
  - [x] normal moves
  - [x] promotion
  - [x] en passant
  - [x] castling
- [ ] legal action
  - [x] can move
  - [x] between 
  - [x] init pawn
  - [ ] en passant
  - [ ] castling
  - [x] underpromotion
  - [x] suicide
- [ ] observation
- [ ] terminal condition
  - [ ] checkmate
  - [ ] stalemate
  - [ ] draw
    - [ ] 50 moves

### speed up

- [ ] cache legal action map
- [ ] compute mask for 1792 normal moves
- [ ] ~~compute mask for 16 dir * 64 target (like shogi)~~
- [ ] compute mask for 12 my piece pos * 64 targets

### CODE REVIEW

**Comment by youyou-ku** [comment]
indexの振り方：
クイーンサイドの自陣側から上に向かって順に振っていく。白番（先手）なら1a(左下白ルーク), 2a（左端白ポーン）, ... ,7h(右端黒ポーン), 8h(右上黒ルーク)の順。黒番は8a(左上黒ルーク), 7a, ..., 2h, 1h(右下白ルーク) の順になる
行動空間：
73 * position(from) + directionでactionが生成される

**Comment by youyou-ku** [comment]
direction:
0~8 under promotion: 3で割った商が0ならrook、1ならbishop、2ならKnightにプロモーション
余りが0ならまっすぐ、1なら右側、2なら左側にうごく
9~64 queen move: 下、上、左、右、左下、右上、左上、右下の順
9：下に7マス　10：下に6マス　…　15：下に1マス　16：上に1マス　…　22：上に7マス
他の向き（左右23~36 斜め上37~50 斜め下51~64）も同様
65~72knight move:

**Comment by sotetsuk** [comment]
From #699 

```py
    # en_passant
    fen = "7k/7p/8/6P1/8/8/8/K7 b - - 0 1"
    s = _from_fen(fen)
    # BPawn: h7 -> h5
    action = 4178
    s, _, _ = step(s, action)
    assert _legal_actions(s).sum() == 5
    # black
    fen = "7k/8/8/8/1p6/8/P7/K7 w - - 0 1"
    s = _from_fen(fen)
    # WPawn: a2 -> a4
    action = 90
    s, _, _ = step(s, action)
    assert _legal_actions(s).sum() == 5
    # promotion
    # TODO: index error
    fen = "b1r4k/1P6/8/8/8/8/8/K7 b - - 0 1"
    s = _from_fen(fen)
    # BKing: h8 -> h7
    action = 4104
    s, _, _ = step(s, action)
    assert _legal_actions(s).sum() == 15
    # black
    fen = "7k/8/8/8/8/8/6p1/K4R1B w - - 0 1"
    s = _from_fen(fen)
    # WKing: a1 -> a2
    action = 16
    s, _, _ = step(s, action)
    assert _legal_actions(s).sum() == 15
    # castling(cannot)
    fen = "7k/8/7p/8/8/8/8/R3K2R b - - 0 1"
    s = _from_fen(fen)
    # BKing: h8 -> h7
    action = 4104
    s, _, _ = step(s, action)
    assert _legal_actions(s).sum() == 22
    # black
    fen = "r2k3r/8/8/8/8/P7/8/K7 w - - 0 1"
    s = _from_fen(fen)
    # WKing: a1 -> a2
    action = 16
    s, _, _ = step(s, action)
    assert _legal_actions(s).sum() == 22
    # castling(cannot)
    fen = "7k/8/7p/8/8/8/8/RN2K1NR b KQ - 0 1"
    s = _from_fen(fen)
    # BKing: h8 -> h7
    action = 4104
    s, _, _ = step(s, action)
    assert _legal_actions(s).sum() == 23
    # black
    fen = "rn1k2nr/8/8/8/8/P7/8/K7 w kq - 0 1"
    s = _from_fen(fen)
    # WKing: a1 -> a2
    action = 16
    s, _, _ = step(s, action)
    assert _legal_actions(s).sum() == 23
    # castling(cannot)
    fen = "1r4rk/8/7p/8/8/8/8/R3K2R b KQ - 0 1"
    s = _from_fen(fen)
    # BKing: h8 -> h7
    action = 4104
    s, _, _ = step(s, action)
    assert _legal_actions(s).sum() == 23
    # black
    # TODO: fix black castling
    fen = "r2k3r/8/8/8/8/P7/8/KR4R1 w kq - 0 1"
    s = _from_fen(fen)
    # WKing: a1 -> a2
    action = 16
    s, _, _ = step(s, action)
    assert _legal_actions(s).sum() == 23
    # castling
    fen = "7k/8/7p/8/8/8/8/R3K2R b KQ - 0 1"
    s = _from_fen(fen)
    # BKing: h8 -> h7
    action = 4104
    s, _, _ = step(s, action)
    assert _legal_actions(s).sum() == 24
    # black
    # TODO: fix black castling
    fen = "r2k3r/8/8/8/8/P7/8/K7 w kq - 0 1"
    s = _from_fen(fen)
    # WKing: a1 -> a2
    action = 16
    s, _, _ = step(s, action)
    assert _legal_actions(s).sum() == 24
    # checked
    # TODO: filter leave-check
    fen = "7k/8/8/8/8/1p6/P7/K7 b - - 0 1"
    s = _from_fen(fen)
    # BPawn: b3 -> b2
    action = 965
    s, _, _ = step(s, action)
    assert _legal_actions(s).sum() == 2
    # suicide
    # TODO: filter suicide-move
    fen = "7k/8/8/8/8/8/7q/K7 b - - 0 1"
    s = _from_fen(fen)
    # BKing: h8 -> h7
    action = 4104
    s, _, _ = step(s, action)
    assert _legal_actions(s).sum() == 1
    # discovered check
    # TODO: filter leave-check
    fen = "7k/8/8/8/3b4/2p5/8/KR6 b - - 0 1"
    s = _from_fen(fen)
    # BPawn: c3 -> c2
    action = 1549
    s, _, _ = step(s, action)
    assert _legal_actions(s).sum() == 2
    # check with pin
    # TODO: filter leave-check
    fen = "7k/8/8/8/3b4/2r5/8/KR6 b - - 0 1"
    s = _from_fen(fen)
    # BRook: c3 -> c1
    action = 1550
    s, _, _ = step(s, action)
    assert _legal_actions(s).sum() == 1
    # double check
    # TODO: filter leave-check
    fen = "7k/8/8/8/3b4/2r5/8/K7 b - - 0 1"
    s = _from_fen(fen)
    # BRook: c3 -> c1
    action = 1550
    s, _, _ = step(s, action)
    assert _legal_actions(s).sum() == 1
    # pin
    # TODO: filter suicide
    fen = "7k/8/8/8/8/8/7r/KR6 b - - 0 1"
    s = _from_fen(fen)
    # BRook: h2 -> h1
    action = 4542
    s, _, _ = step(s, action)
    assert _legal_actions(s).sum() == 8
    # double pin
    # TODO: filter suicide
    fen = "7k/6b1/5r2/8/3B4/8/8/KR6 b - - 0 1"
    s = _from_fen(fen)
    # BRook: f6 -> f1
    action = 3086
    s, _, _ = step(s, action)
    assert _legal_actions(s).sum() == 11
```

**Comment by sotetsuk** [comment]
アンパッサンとキャスリング以外はだいたい終わったか。ランダムプレイもそれっぽく大体動く。

**Comment by sotetsuk** [comment]
### 実行速度検証

```py
import time
import pgx
import jax
from tqdm import tqdm
import jax.numpy as jnp
from pgx.experimental.utils import act_randomly
from pgx.experimental.wrappers import auto_reset
from pgx.chess import Chess

env = Chess(max_termination_steps=1000)


init = jax.jit(jax.vmap(env.init))
step = jax.jit(jax.vmap(auto_reset(env.step, env.init)))
act_randomly = jax.jit(act_randomly)
N = 10
key = jax.random.PRNGKey(0)
key, subkey = jax.random.split(key)
keys = jax.random.split(subkey, N)
state: pgx.State = init(keys)
key, subkey = jax.random.split(key)
action = act_randomly(subkey, state)
state = step(state, action)

N = 10
key = jax.random.PRNGKey(0)
key, subkey = jax.random.split(key)
keys = jax.random.split(subkey, N)
state: pgx.State = init(keys)
i = 0
st = time.time()
for i in tqdm(range(1000)):
    key, subkey = jax.random.split(key)
    action = act_randomly(subkey, state)
    state = step(state, action)
    i += 1
et = time.time()
print(et - st)
```

現状 `21.604450941085815`

**Comment by sotetsuk** [comment]
CIが通れば一旦ここまででマージする
